roles/openshift_health_checker/library/aos_version.py

```diff
@@ -1,91 +1,189 @@
+going to work out right.
+
+TODO: fail gracefully on non-yum systems (dnf in Fedora)
 '''
 
 import yum  # pylint: disable=import-error
 
 from ansible.module_utils.basic import AnsibleModule
 
 
-def main():  # pylint: disable=missing-docstring,too-many-branches
+class AosVersionException(Exception):
+    '''Base exception class for package version problems'''
+    def __init__(self, message, problem_pkgs=None):
+        Exception.__init__(self, message)
+        self.problem_pkgs = problem_pkgs
+
+
+def main():
+    '''Entrypoint for this Ansible module'''
     module = AnsibleModule(
         argument_spec=dict(
-            prefix=dict(required=True),  # atomic-openshift, origin, ...
-            version=dict(required=True),
+            requested_openshift_release=dict(type="str", default=''),
+            openshift_deployment_type=dict(required=True),
+            rpm_prefix=dict(required=True),  # atomic-openshift, origin, ...?
         ),
         supports_check_mode=True
     )
 
-    def bail(error):  # pylint: disable=missing-docstring
-        module.fail_json(msg=error)
+    # determine the packages we will look for
+    rpm_prefix = module.params['rpm_prefix']
+    if not rpm_prefix:
+        module.fail_json(msg="rpm_prefix must not be empty")
+    expected_pkgs = set([
+        rpm_prefix,
+        rpm_prefix + '-master',
+        rpm_prefix + '-node',
+    ])
 
-    rpm_prefix = module.params['prefix']
+    # determine what level of precision the user specified for the openshift version.
+    # should look like a version string with possibly many segments e.g. "3.4.1":
+    requested_openshift_release = str(module.params.get('requested_openshift_release', ''))
+
+    # get the list of packages available and complain if anything is wrong
+    try:
+        pkgs = _retrieve_available_packages(expected_pkgs)
+        if requested_openshift_release:
+            _check_precise_version_found(pkgs, expected_pkgs, requested_openshift_release)
+            _check_higher_version_found(pkgs, expected_pkgs, requested_openshift_release)
+        if module.params['openshift_deployment_type'] != 'origin':
+            _check_multi_minor_release(pkgs, expected_pkgs)
+    except AosVersionException as excinfo:
+        module.fail_json(msg=str(excinfo))
+    module.exit_json(changed=False)
 
-    if not rpm_prefix:
-        bail("prefix must not be empty")
 
+def _retrieve_available_packages(expected_pkgs):
+    # search for package versions available for openshift pkgs
     yb = yum.YumBase()  # pylint: disable=invalid-name
-    yb.conf.disable_excludes = ["all"]  # assume the openshift excluder will be managed, ignore current state
 
-    # search for package versions available for aos pkgs
-    expected_pkgs = [
-        rpm_prefix,
-        rpm_prefix + '-master',
-        rpm_prefix + '-node',
-    ]
+    # The openshift excluder prevents unintended updates to openshift
```

Comment:
This seems to be the best I can find:

https://wiki.centos.org/SpecialInterestGroup/PaaS/OpenShift-Origin-Control-Updates

Our docs don't say much about them, and may say less in the future as you're not really supposed to have to think about them. They're actually totally simple. It's just the use cases that are complex.